File: app/routers/credits.py
from fastapi import APIRouter, Depends
from app.models.schemas import CreditsResponse, DeductCreditsRequest, DeductCreditsResponse
from app.services.supabase_service import get_supabase_service, SupabaseService

router = APIRouter(prefix="/credits", tags=["credits"])


@router.get(
    "",
    response_model=CreditsResponse,
    summary="Get User Credits",
    description="Get the current user's credit balance"
)
async def get_credits(
    supabase_service: SupabaseService = Depends(get_supabase_service)
):
    """
    Get user's remaining credits for text, image, and video generation.
    """
    credits = await supabase_service.get_user_credits("demo-user")
    return CreditsResponse(
        textCredits=credits.get("text_credits", 150),
        imageCredits=credits.get("image_credits", 25),
        videoCredits=credits.get("video_credits", 10),
        totalCredits=credits.get("text_credits", 150) + credits.get(
            "image_credits", 25) + credits.get("video_credits", 10),
        plan=credits.get("plan", "free")
    )


@router.post(
    "/deduct",
    response_model=DeductCreditsResponse,
    summary="Deduct Credits",
    description="Deduct credits after content generation"
)
async def deduct_credit(
    request: DeductCreditsRequest,
    supabase_service: SupabaseService = Depends(get_supabase_service)
):
    """
    Deduct credits from user's balance.

    - **creditType**: "text", "image", or "video"
    - **amount**: Number of credits to deduct (default: 1)
    """
    credit_map = {
        "text": "text_credits",
        "image": "image_credits",
        "video": "video_credits"
    }

    if request.creditType not in credit_map:
        return DeductCreditsResponse(
            success=False,
            creditsRemaining=0,
            message="Invalid credit type"
        )

    success = await supabase_service.deduct_credit("demo-user", credit_map[request.creditType])
    credits = await supabase_service.get_user_credits("demo-user")

    return DeductCreditsResponse(
        success=success,
        creditsRemaining=credits.get(credit_map[request.creditType], 0),
        message="Credits deducted successfully" if success else "Insufficient credits"
    )

File: routes/credits.js
import express from 'express';
import 'express-async-errors';
import { getSupabaseService } from '../services/supabaseService.js';

const router = express.Router();

// GET /credits - Get user credits
router.get('/', async (req, res) => {
    try {
        const supabaseService = getSupabaseService();
        
        // TODO: Get actual user_id from authentication
        const userId = '00000000-0000-0000-0000-000000000000';
        
        const credits = await supabaseService.getUserCredits(userId);
        
        res.json({
            textCredits: credits.text || 150,
            imageCredits: credits.image || 25,
            videoCredits: credits.video || 10,
            totalCredits: (credits.text || 150) + (credits.image || 25) + (credits.video || 10),
            plan: 'free'
        });
    } catch (error) {
        console.error(`Credits fetch error: ${error.message}`);
        res.status(500).json({
            success: false,
            error: error.message || 'Failed to fetch credits'
        });
    }
});

// POST /credits/deduct - Deduct credits
router.post('/deduct', async (req, res) => {
    try {
        const { creditType, amount = 1 } = req.body;
        
        if (!creditType || !['text', 'image', 'video'].includes(creditType)) {
            return res.status(400).json({
                success: false,
                message: 'Invalid credit type. Must be text, image, or video.'
            });
        }
        
        const supabaseService = getSupabaseService();
        
        // TODO: Get actual user_id from authentication
        const userId = '00000000-0000-0000-0000-000000000000';
        
        const success = await supabaseService.deductCredit(userId, creditType);
        
        if (success) {
            const credits = await supabaseService.getUserCredits(userId);
            res.json({
                success: true,
                creditsRemaining: credits[creditType] || 0,
                message: `${amount} ${creditType} credit(s) deducted successfully`
            });
        } else {
            res.status(402).json({
                success: false,
                creditsRemaining: 0,
                message: 'Insufficient credits'
            });
        }
    } catch (error) {
        console.error(`Credits deduction error: ${error.message}`);
        res.status(500).json({
            success: false,
            error: error.message || 'Failed to deduct credits'
        });
    }
});

export default router;
